hugo/assets/js/species_cards.js

```diff
@@ -0,0 +1,284 @@
+/*
+Handles searching/filtering, ordering and pagination of the species cards on the home page.
+This script works with 'index.html' inside the hugo/layouts folder.
+
+On load:
+1. Species attributes needed for searching fetched from the <template> cards built for each species.
+2. Determine if pagination exists (decided by Hugo based on number of species and 'maxSpeciesPerPage' in 'hugo/config.yaml').
+3. Run the render function to display the species cards that should be shown.
+
+Once loaded:
+- Event listners created for search box, dropdown and pagination (if pagination exists).
+- Each event listner calls the main render function (renderSpeciesCards) alongside some others dependant on the event.
+
+Notes:
+- This code can handle both pagination and no pagination (i.e. not enough species to warrant it). So can the Playwright tests assocaited with this page.
+*/
+
+const SORT_OPTIONS = {
+    LAST_UPDATED: 'lastUpdated',
+    A_TO_Z: 'alphabet',
+    Z_TO_A: 'revAlphabet'
+};
+
+const paginationItems = document.querySelectorAll('.pagination .page-item');
+const noResultsCard = document.getElementById('no-filtered-card');
+const PAGINATION_EXISTS = paginationItems.length > 0; // if not enough species, there will be no pagination yet.
+const CARDS_PER_PAGE = parseInt(document.getElementById('card-container').dataset.numbCardsPerPage);
+
+let state = {
+    speciesData: [], // updated by reading from the <template>s in the HTML
+    currentPage: 1,
+    sortOrder: SORT_OPTIONS.LAST_UPDATED,
+    searchText: '',
+    numbMatches: 0 // updated when search ran.
+};
+
+
+/**
+ * Runs on page load.
+ * Gets the info needed to search and order the species cards from each <template> card.
+ * Runs the render function to display the species cards.
+ */
+function prepareSpeciesPage() {
+    const templates = document.querySelectorAll('template[id^="card-"]');
+    templates.forEach(template => {
+        const title = template.content.querySelector('.science-name').textContent.toLowerCase();
+        const subtitle = template.content.querySelector('.common-name').textContent.toLowerCase();
+        const species = {
+            id: template.id,
+            speciesText: title + ' ' + subtitle, // used for searching
+            last_updated: template.content.querySelector('.card-text').textContent.replace('Last updated: ', ''),
+        };
+        state.speciesData.push(species);
+    });
+    state.numbMatches = templates.length
+
+    renderSpeciesCards();
+}
+
+
+// Sorting functions below
+
+function stringToDate(str) {
```

Comment:
Here is another weak point where changing the date format in Hugo would break the search. The very hypothetical day we decide to localize the site for US viewers for example. To be contrasted with sorting based on a `date` JSON field which would be unambiguously serialized.

But these are again lofty design considerations, I would be fine to stick with this if we forsake the JSON approach.

Reply:
Fair point, I added a Playwright test to specifically catch a change in these date formats now. Not a perfect solution perhaps but hopefully enough to make one consider. Fixing the test hopefully results in reading the test docstring which would let the dev know to change the JS too. commit 267ac4e593bd12d801d7317ecf6526227867cb04